docs/HowToFAQ/FAQLoopModifies.md

```diff
@@ -0,0 +1,58 @@
+---
+title: If I have an assertion about a class and a loop that doesn't mention (read, modify)  the class, why does dafny fail to establish the assertion after the loop?
+---
+
+## Question
+
+If I have an assertion about a class and a loop that doesn't mention (read, modify)  the class, why does dafny fail to establish the assertion after the loop?
+
+## Answer
+
+TODO - This may need adjustment of the modifies/assignmetn targets discussion
+
+The short answer is that you need an appropriate combination of modifies clause and loop invariants to prove assertions about the end result of a loop.
+
+In Dafny's way of reasoning about a loop (which is typical of verifier systems), the loop invariant is the only thing you can assume at the top of every iteration. In other words, at the top of each loop iteration, it is as if all variables have arbitrary values, modulo the loop invariant. Sometimes, the word _havoc_ is used to describe this: the verifier havocs all variables, that is, gives all variables arbitrary values, and then assumes the loop invariant.
+
+If that’s all the verifier did, life would be a pain, because you’d have to write a loop invariant about all variables in your program, even those that have nothing to do with the loop.
+Verifier actually does better. Instead of havocing all variables, it suffices to havoc the assignment targets within the body of the loop, including anything that might be modified by methods called in the loop body. That is, the verifier uses a simple syntactic scan of the loop body to see which variables may possibly be assigned, and then it havocs only those variables. This saves users from having to write loop invariants about all other variables. Only invariants about variables
```

Comment:
And then here again we say something that is wrong (havocs only those variables), and then refine again later

Reply:
As above